### PR TITLE
[IMP] website : shapes on images

### PR DESCRIPTION
**Pull request by detrouxdev**
With this improvement, user can now choose between a set of shapes to clip around an image. This allows for more dynamic representation of images.

This pull request also contains a tool that allows for a faster conversion of SVGs coming from Illustrator into usable shapes inside this feature.

### CODE REVIEW

**Comment by detrouxdev** [comment]
Original PR : https://github.com/odoo/odoo/pull/68142

**Comment by detrouxdev** [comment]
@sdegueldre we decided that we would merge the feature before working on animations. Could you have a look at the code in its current state? One thing I'd like your comment on is the setTimeout on the select Option to prevent reset flickering when switching from one shape to another. I was told you found a better way than doing it with a setTImeout. Thanks 😄

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Cool feature :)

- Can you add inline comments to explain what blocks of code are actually doing so I can properly review? Not each line of code of course but some methods could use one or two comments.

- Shapes are saved with a .jpg URL but using a svg mimetype... this has to be fixed. It is strange, I am sure there will be problems on some browsers and actually this is a problem as soon as you DL an image -> your OS will not like it (at least mine does not)

- I don't remember if we talked about that with Sam or not but background shapes have a class for their default config so that colors change if the theme color are changed (if the shape color were not customized by the user)... this is not the case with image shape so it is really not consistent. No way to solve that here I guess?

- Any reason to prevent adding shapes around SVG images? The optimizations were disabled but is it required to disable shapes too?

File: addons/web_editor/static/image_shapes/convert-to-percentages.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
+// marks which axis each parameter of a command belongs to, as well as whether
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/static/image_shapes/convert-to-percentages.js:1]: This hurt my feelings. I think the indentation is random because I switched from atom to vscode while writing this and my tab-width setting wasn't the same :/ I'd love to turn on an eslint warning for indentation but because the code inside odoo modules is not indented I haven't been able to, maybe with the new modules I could try again. Arthur picked it up half way with the indentation already messed up I guess.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/image_shapes/convert-to-percentages.js:1]: > This hurt my feelings.

But the indentation hurt my eyes :cry: Anyways, that was a real question, did not look at the code at all :wink: @detrouxdev Can you reindent the file correctly? Thanks :+1:

File: addons/web_editor/views/snippets.xml [hunk 2]
```diff
@@ -380,6 +380,34 @@
     <div data-js="ImageOptimize"
          data-selector="img">
         <t t-call="web_editor.snippet_options_image_optimization_widgets"/>
+        <we-row string="Shapes">
+            <we-select-pager data-name="shape">
+                <we-select-page string="Solids">
+                    <we-button data-shape="web_editor/blob_1_solid_rd" data-select-label="Blob Solid 1">Blob Solid 1</we-button>
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/views/snippets.xml:386]: I was uncertain about the look of the selector as well but SBU seemed ok with it. Would something like this be more suitable? (only did the first 2 for the example) I'm unsure of the way I could make this look right. Do you have any advice?

Screenshot 2021-05-05 at 10 29 26

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -4453,11 +4472,138 @@ registry.ImageOptimize = ImageHandlerOption.extend({
...
         return this._super(...arguments);
     },
+    /**
+     * @override
+     */
+    cleanForSave() {
+        if (this._getImg().dataset.shape) {
+            const img = this._getImg();
+            img.dataset.originalMimetype = img.dataset.mimetype;
+            img.dataset.mimetype = 'image/svg+xml';
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/static/src/js/editor/snippets.options.js:4482]: In order to preserve Image Optimisation features we keep the img.dataset.mimetype as its original one during edition.
Once it's going to be saved however, the server needs to know its an SVG and not a JPEG or PNG for it to be displayed properly again.

File: addons/web_editor/views/snippets.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/views/snippets.xml:386]: What about a grid instead of a list? Like 2 or 3 square per line

[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/views/snippets.xml:386]: That would definitely look better! Is there a widget or a way to do it that already exists? Or should I make one?

[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/web_editor/views/snippets.xml]: I used `<we-row>` inside the pager, would it be ok?
Screenshot 2021-05-06 at 13 38 54

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/views/snippets.xml]: Not a fan of using we-row (not there for that and you have to declare the row by yourself). For me it should be automatic with a unique class added on the we-select declaration (something like "o_we_select_grid"). And I am sure you can make a more beautiful result than the one you show here which looks like in-dev stuff :wink: (more consistent horizontal/vertical spacing, padding around the shape, same size for all squares, etc).

**Comment by detrouxdev** [comment]
> I don't remember if we talked about that with Sam or not but background shapes have a class for their default config so that colors change if the theme color are changed (if the shape color were not customized by the user)... this is not the case with image shape so it is really not consistent. No way to solve that here I guess?

Maybe @sdegueldre can comment on this. I remember us discussing of the way to handle colors and it's different than the one we use for bg shapes. I replace the colors when we first create the image and it's then saved server side, loaded in an img tag. I cannot use the theme's CSS variable inside the svg because of it.

I'm unsure as to why we cannot do it the same way we handle BG shapes but I think the reason is : get the base svg, modify the colors and place the image inside of it in base64. We then save this new svg on disk as an attachement. I think BG shapes are called server side and a controller changes their color values before sending it to the client. Making it different than loading an attachement I suppose?

**Comment by sdegueldre** [comment]
> I don't remember if we talked about that with Sam or not but background shapes have a class for their default config so that colors change if the theme color are changed (if the shape color were not customized by the user)... this is not the case with image shape so it is really not consistent. No way to solve that here I guess?

Not that I know of. We can't change the src of an  depending on a class in a cross-browser way, even if we could, the image needs to be inside the svg for the browser to load it, so we would need one class and attachment per image+shape combination. Since the bg-shapes don't contain another image, having a class for each shape in its default configuration makes sense, but not for image-shapes since they only contain a grey blob in their default configuration.

**Comment by qsm-odoo** [comment]
Yes, let's forget about automatic-color-shapes.

**Comment by qsm-odoo** [comment]
@detrouxdev 

> I think BG shapes are called server side and a controller changes their color values before sending it to the client
No, it's just that default bg shapes are added by a CSS class whose background image's src is dynamically built with colors that the scss know :wink:

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
Two small comments then we can send to testing :+1:

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js [hunk 4]
```diff
@@ -61,7 +61,9 @@ var MassMailingFieldHtml = FieldHtml.extend({
         }
 
         var $editable = this.wysiwyg.getEditable();
-
+        if (this.wysiwyg.snippetsMenu) {
+            await this.wysiwyg.snippetsMenu.cleanForSave();
+        }
```
Comment (qsm-odoo) [inline addons/mass_mailing/static/src/js/mass_mailing_widget.js:66]: This entire commit seems like an important fix to do in stable versions if it is actually needed. Can you verify and move it to another PR if this is the case?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 5]
```diff
@@ -377,7 +377,7 @@ const Wysiwyg = Widget.extend({
                         res_model: resModel,
                         res_id: parseInt(resId),
                         data: (isBackground ? el.dataset.bgSrc : el.getAttribute('src')).split(',')[1],
-                        mimetype: el.dataset.mimetype,
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: I find this fix really strange, the same way I found Arthur's implementation of mimetype in the `cleanForSave` strange too. For me, the data-mimetype should directly be set to svg by the option which actually add the shape/convert to SVG. And the original mimetype be mentioned directly too and used where it is actually needed.

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 6]
```diff
@@ -819,8 +819,6 @@ body.editor_enable.editor_has_snippets {
         // Buttons
         we-button.o_we_user_value_widget {
             > div {
-                // Needed otherwise cannot work because of flex display
-                @include o-text-overflow(block);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss:823]: Kinda scare by the fact you remove a line which is explained as "Needed otherwise..." without explanation and that your commit message says "Needs more testing" :grimacing: 

With this line removed, you lose the "..." effect in button whose text is too long so that can't be the right solution here

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 7]
```diff
@@ -594,7 +594,7 @@ const Wysiwyg = Widget.extend({
     openMediaDialog(params = {}) {
         const sel = this.odooEditor.document.getSelection();
         const fontawesomeIcon = getInSelection(this.odooEditor.document, '.fa');
-        if (fontawesomeIcon && sel.toString().trim() === "") {
+        if (fontawesomeIcon && sel.toString().trim() === "" && !$(sel.anchorNode).hasClass('o_we_user_value_widget')) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:597]: Can't be the right fix from what I understand from your commit explanation. Is this bug also there in 14.0 or is this since the new editor?

File: odoo/tools/mail.py [hunk 8]
```diff
@@ -40,7 +40,8 @@
      'data-o-mail-quote',  # quote detection
      'data-oe-model', 'data-oe-id', 'data-oe-field', 'data-oe-type', 'data-oe-expression', 'data-oe-translation-id', 'data-oe-nodeid',
      'data-publish', 'data-id', 'data-res_id', 'data-interval', 'data-member_id', 'data-scroll-background-ratio', 'data-view-id',
-     'data-class', 'data-mimetype',
+     'data-class', 'data-mimetype', 'data-shape', 'data-original-src', 'data-file-name', 'data-shape-colors',
+     'data-original-mimetype', 'data-original-id', 'data-resize-width',
```
Comment (qsm-odoo) [inline odoo/tools/mail.py:44]: Aren't some of those attributes already used in stable version? Should not this fix target those stable vesions too?

File: addons/web_editor/static/src/js/common/utils.js [hunk 9]
```diff
@@ -185,13 +185,28 @@ function _computeColorClasses(colorNames, prefix = 'bg-') {
...
+        // There could be a better way of doing it.
+        if (window.length > 1) {
+            for (let i = 0; i < window.length; i++) {
+                const iframeWindow = window[i];
+                if (iframeWindow.frameElement && iframeWindow.frameElement.classList.contains('wysiwyg_iframe')) {
+                    const doc = iframeWindow.document;
+                    const tempEl = doc.body.appendChild(doc.createElement('div'));
+                    tempEl.className = `bg-${key}`;
+                    iframeValue = iframeWindow.getComputedStyle(tempEl).getPropertyValue("background-color").trim();
+                    tempEl.parentNode.removeChild(tempEl);
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/common/utils.js:200]: Can't we do this properly by adding the required CSS variable on the iframe HTML when needed and reading values from those like in the website case?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (xO-Tx) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: I Tested both cases and it seems that code was implemented this way to set the right mimetype if image changed...
Let's suppose that we have a JPG image and we moved the code updating data-mimetype to option's method :

- After applying a shape on it, if we replace the image by a PNG image, we'll get as a result an image with
dataset = {..., mimetype: 'image/png', originalMimetype: 'image/jpeg'} (an update to SVG is always needed)

I think we can keep it this way or add it in options method in a way to keep the real mimetype when image changed by user. [Added a separate commit for the second implementation(*)].

(*) : I added commits by "push" as requested in task PAD with a "v-i" to track changes of a single commit.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (xO-Tx) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: While debugging this code, the only difference I found between document in website and iframe is that current **selection** in iframe editor starts from the option's button itself and then considered as an update of `fontawesomeIcon`

14 3-editor-target-selection

14 3-image-target-error

As a consequence, target node will be replaced in this code :

```
if (fontawesomeIcon && sel.toString().trim() === "") {
    params.node = $(fontawesomeIcon);
    ...
}
```

Which was added on 14.3 as a [fix for icon replacement](https://github.com/odoo/odoo/pull/70118/commits/756783665e07d59bef57880c0dce72ff259b6a60).

That's why I tried to exclude the 'o_we_user_value_widget' button (I reverted this commit since it should target saas-14.3 anyway).

File: addons/web_editor/static/src/js/common/utils.js (hunk 9 above)
Comment (xO-Tx) [reply addons/web_editor/static/src/js/common/utils.js:200]: I checked this possibility but the situation here is that some default theme color classes (e.g. '.bg-o-color-4') are overridden by CSS using a template related class (e.g. '.o_resto_theme') to set new values on them, so I don't really see how we can do it in a clean way :thinking:

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_widget.js:66]: I think this is safe to do in stable versions (ignore < 13.0) to handle potential custo or community modules.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: I am not sure to understand everything that you are saying here. My point was to adapt all related code so that any option changing the image should directly lead to a consistent `mimetype` + `originalMimetype`. So in your example: ` {..., mimetype: 'image/png', originalMimetype: 'image/jpeg'}` should simply not be possible -> you change the image, both values are changed, you add a shape, only mimetype is changed, etc etc.

Did not look at your code update yet, is this what you did?

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 7 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: I opened an editor feedback for this: https://www.odoo.com/web#id=2567806&action=333&active_id=1695&model=project.task&view_type=form&cids=1&menu_id=4720 so you can ignore it

File: addons/web_editor/static/src/js/common/utils.js (hunk 9 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/common/utils.js:200]: Those damn mailing templates :/ But so this is not really related to the fact that this is inside an iframe so this patch feels really bad :grimacing: What about overriding the method which calls this `_getCSSVariableValue` when inside the mailing editor instead? That would be the proper way to do it I think.

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js (hunk 5 above)
Comment (xO-Tx) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js]: Actually, it's the same idea, the `mimetype`/`orignalMimetype` keep the right values when image changed & option applied.
I added an update on the code to handle the case where shape removed (`data-mimetype` should get the original
value and `data-original-mimetype` should be removed).

File: addons/web_editor/static/src/js/common/utils.js (hunk 9 above)
Comment (xO-Tx) [reply addons/web_editor/static/src/js/common/utils.js]: Actually the `_getCSSVariableValue` is used in option's methods, so I updated the code to override its behaviour inside mass mailing iframe.

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js (hunk 4 above)
Comment (xO-Tx) [reply addons/mass_mailing/static/src/js/mass_mailing_widget.js:66]: I checked the code but it seems that in [14.0 - 14.2] we can't get the `this.wysiwyg.snippetsMenu` object (to run its `cleanForSave` method) in fieldHtml code. Is this intended to be this way ?

File: addons/web_editor/static/src/js/common/utils.js (hunk 9 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/common/utils.js]: I understand that, my point was to override the option's behavior to not use `_getCSSVariableValue` (since you don't want a CSS variable there)

File: addons/mass_mailing/static/src/js/mass_mailing_widget.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/mass_mailing/static/src/js/mass_mailing_widget.js:66]: Ok, if code different in the old editor, we can care about fixing that in stable versions later.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -4905,6 +4902,27 @@ registry.ImageOptimize = ImageHandlerOption.extend({
...
+        if (doc && doc.querySelector('.o_mass_mailing_iframe')) {
+            const tempEl = doc.body.appendChild(doc.createElement('div'));
+            tempEl.className = `bg-${color}`;
+            const colorValue = window.getComputedStyle(tempEl).getPropertyValue("background-color").trim();
+            tempEl.parentNode.removeChild(tempEl);
+            return ColorpickerWidget.normalizeCSSColor(colorValue).replace(/"/g, "'");
+        } else {
+            return weUtils.getCSSVariableValue(color);
+        }
+    },
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:4936]: You mention a mass mailing class inside of the web_editor app -> not ok. But this is going in the right direction: what you have to do is define that `_getCSSColorValue` here with only the normal "website" case -> reading the CSS variable. Then, in mass mailing (something mass_mailing_editor.js), you can **include** this JS class to change the `_getCSSColorValue` method behavior.

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 11]
```diff
@@ -4701,14 +4689,22 @@ registry.ImageOptimize = ImageHandlerOption.extend({
...
         } else {
             // Re-applying the modifications and deleting the shapes
             img.src = await applyModifications(img);
             delete img.dataset.shape;
             delete img.dataset.shapeColors;
             delete img.dataset.fileName;
+            if (saveData) {
+                img.dataset.mimetype = img.dataset.originalMimetype;
+                delete img.dataset.originalMimetype;
+            }
```
[the remark this answers is not in the document]
Comment (xO-Tx) [reply addons/web_editor/static/src/js/editor/snippets.options.js:4647]: 1- Let's take the example of `willStart`, a saved PNG image with a shape will have `dataset = {..., mimetype: 'image/svg+xml', originalMimetype: 'image/png'}` (which make sense)

But on `willStart`, we should set the original value (`'image/png'`) on `img.dataset.mimetype` to have image options available **(see ImageHandlerOption > _renderCustomXML)**. Then, we should keep the right mimetype value since the image will be saved by the `saveModifiedImages` method...

The `dataset.mimetype` is also used by `image_processing` > `applyModifications` to apply options on images, so without the code in `willStart`, we won't be able to update correctly images inside shapes.

I updated the code to get the right mimetype for image when updated by options using a `_getImageMimetype` method.
It's better this way and we don't need to add the original mimetype on `willStart` :slightly_smiling_face:

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 12]
```diff
@@ -4669,19 +4669,6 @@ registry.ImageOptimize = ImageHandlerOption.extend({
         this.$target.off('.ImageOptimization');
         return this._super(...arguments);
     },
-    /**
-     * @override
-     */
-    cleanForSave() {
```
[the remark this answers is not in the document]
Comment (xO-Tx) [reply addons/web_editor/static/src/js/editor/snippets.options.js]: Yes, `snippetEditors` are destroyed on mass mailing if form saving fails (e.g. save the form without a required field...).
Actually that's the reason why changes in `cleanForSave` are not applied on mass mailing in this flow...
I will add a separate task for that :+1:

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 13]
```diff
@@ -1016,19 +1016,23 @@ body.editor_enable.editor_has_snippets {
...
-                display: flex;
-                flex-wrap: wrap;
-                we-button {
-                    width: percentage(1 / 3);
-                    padding: 0;
-                    justify-content: center;
-                    img {
-                        padding: $o-we-sidebar-content-field-spacing;
+            we-select-page {
+                max-height: 62px * 2.7;
```
[the remark this answers is not in the document]
Comment (xO-Tx) [reply addons/web_editor/static/src/scss/wysiwyg_snippets.scss]: Yes, it depends on the height of the image inside `we-button`... I moved the code to CSS, I think it's better this way :+1:

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+
